impeller/geometry/matrix.h

```diff
@@ -357,14 +357,18 @@ struct Matrix {
   }
 
   constexpr Vector3 operator*(const Vector3& v) const {
+    Scalar w = v.x * m[3] + v.y * m[7] + v.z * m[11] + m[15];
```

Comment:
This will need a guard against zero. I assume you realize this because of your comment, just making sure I don't lose track when I look at this again next time :) 

Ditto below.

Reply:
On the fence about whether it should be inf or 0 at the perspective singularity. TBH accepting UB here seems fine as well.

Reply:
Eh, actually it's not the perspective singularity, it's just a poorly constructed matrix that makes no sense because it has no position information. I'm starting to lean more towards accepting UB the more I think about this.

Reply:
Where would it make sense to document the UB? Is there somewhere we could put that in `dart:ui`?

Reply:
We could document the behavior on any surface that appends the canvas CTM.  I looked at how Skia does it and they go with the 0 approach: https://github.com/google/skia/blob/13a15693a1834c9d946807f5dd6867932e1d11e8/src/core/SkMatrix.cpp#L982-L987

I'm gonna switch to that for consistency and add a note inline that it would be reasonable to allow UB in the future when optimizing

Reply:
The thing to avoid for Flutter is a hard application crash because a developer hot reloaded into a bad state.